Pad rev 4 encryption keys to be >= 16 bytes (issue 19484)

rossj opened the pull request:
This PR fixes issue https://github.com/mozilla/pdf.js/issues/19484 by extending too-short keys to 16 bytes (padded with 0s) for revision 4 encryption. This key-extension behavior seems undocumented, but is necessary to properly decrypt string & stream data of problematic files found in the wild (supposedly generated by InDesign).

rossj committed:
Fix decryption of R=4, V=4 files with < 16-byte keys by 0-padding - undocumented but matches Acrobat behavior (issue #19484)

Snuffleupagus commented:
- Does this patch pass all existing, i.e. `npx gulp test`, tests locally?

 - Since I believe that you created those PDFs yourself, could we perhaps commit them to repository instead rather than relying on links (since that's always preferable where possible?

 - Can you please add more context to the *commit message*, i.e. what you included in https://github.com/mozilla/pdf.js/pull/19555#issue-2875886668, to make the patch understandable on e.g. the `git` command line as well?

Snuffleupagus commented on src/core/crypto.js:
```diff
@@ -1787,7 +1787,14 @@ class CipherTransformFactory {
 
-    this.encryptionKey = encryptionKey;
+    if (algorithm >= 4 && encryptionKey.length < 16) {
+      // Extend key to 16 byte minimum (undocumented)
+      // Fixes issue19484_1.pdf & issue19484_2.pdf
```
```suggestion
      // Extend key to 16 byte minimum (undocumented),
      // fixes issue19484_1.pdf and issue19484_2.pdf.
```

rossj replied:
Thanks, updated

rossj commented:
> * Does this patch pass all existing, i.e. `npx gulp test`, tests locally?

Mostly, except for:
- issue5939.pdf - cannot download as link is 404
- Repeated failure for `caches image resources at the document/page level, with main-thread copying of complex images (issue 11518) | in chrome | Expected 1160 to be less than 1064.5.` - seems like a timing / perf check, so may be a flaky test
- I have a number of pages that don't match ref in Chrome, even if I generate fresh refs. Looks like minor single-pixel color differences in the diff viewer. Firefox refs all match.

> * Since I believe that you created those PDFs yourself, could we perhaps commit them to repository instead of relying on links (since that's always preferable where possible)?

Updated 

> * Can you please add more context to the _commit message_, i.e. what you included in [Pad rev 4 encryption keys to be >= 16 bytes (issue 19484) #19555 (comment)](https://github.com/mozilla/pdf.js/pull/19555#issue-2875886668), to make the patch understandable on e.g. the `git` command line as well?

Updated

Snuffleupagus commented:
> * issue5939.pdf - cannot download as link is 404

That one *should* have been fixed by a PR earlier today.

> * Repeated failure for `caches image resources at the document/page level, with main-thread copying of complex images (issue 11518) | in chrome | Expected 1160 to be less than 1064.5.` - seems like a timing / perf check, so may be a flaky test

We've seen that occasionally failing on the bots, but as long as Firefox works that fine.

> * I have a number of pages that don't match ref in Chrome, even if I generate fresh refs. Looks like minor single-pixel color differences in the diff viewer. Firefox refs all match.

We actually had so much trouble with *unstable* reference tests in Chrome that we nowadays only test in Firefox for that reason.

Snuffleupagus commented:
/botio unittest

Snuffleupagus commented:
/botio browsertest

Snuffleupagus reviewed (approved):
Hopefully we have enough test-coverage for this part of the code-base, since we're trying to match unspecified Adobe Reader behaviour here (rather than fixing an "actual" bug).
However, I suppose that we have to land this to actually find out if there's any breakage :-)

r=me, thank you.

Snuffleupagus commented:
/botio makeref
